File: app/src/main/java/com/budgetbuddy/ui/home/HomeViewModel.kt
package com.budgetbuddy.ui.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.budgetbuddy.data.local.entities.TransactionEntity
import com.budgetbuddy.data.repository.BudgetRepository
import com.budgetbuddy.data.repository.CategoryRepository
import com.budgetbuddy.data.repository.GoalRepository
import com.budgetbuddy.data.repository.TransactionRepository
import com.budgetbuddy.util.DateUtils
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

data class HomeUiState(
    val balance: Double = 0.0,             // income - expenses (all time)
    val totalIncomeThisMonth: Double = 0.0,
    val totalSpendThisMonth: Double = 0.0,
    val recentTransactions: List<TransactionEntity> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null
)

@HiltViewModel
class HomeViewModel @Inject constructor(
    private val transactionRepository: TransactionRepository,
    private val categoryRepository: CategoryRepository,
    private val budgetRepository: BudgetRepository,
    private val goalRepository: GoalRepository
) : ViewModel() {

    private val _userId = MutableStateFlow("")
    private val _uiState = MutableStateFlow(HomeUiState())
    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()

    val categories = categoryRepository.getAllCategories()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private val _goalsState = MutableStateFlow<List<com.budgetbuddy.data.local.entities.GoalEntity>>(emptyList())
    val goals: StateFlow<List<com.budgetbuddy.data.local.entities.GoalEntity>> = _goalsState.asStateFlow()

    fun init(userId: String) {
        _userId.value = userId
        loadRecentTransactions(userId)
        loadMonthlyTotal(userId)
        loadGoals(userId)
    }

    private fun loadGoals(userId: String) {
        viewModelScope.launch {
            goalRepository.getActiveGoals(userId).collect { goals ->
                _goalsState.value = goals
            }
        }
    }

    private fun loadRecentTransactions(userId: String) {
        viewModelScope.launch {
            transactionRepository.getRecentTransactions(userId, 5).collect { transactions ->
                _uiState.update { it.copy(recentTransactions = transactions) }
            }
        }
    }

    private fun loadMonthlyTotal(userId: String) {
        viewModelScope.launch {
            val start = DateUtils.startOfMonth()
            val end = DateUtils.endOfMonth()
            // Load both income and expense for the month so balance = income - expenses
            val expense = transactionRepository.getTotalExpenseForPeriod(userId, start, end)
            val income  = transactionRepository.getTotalIncomeForPeriod(userId, start, end)
            val balance = income - expense
            _uiState.update {
                it.copy(
                    totalSpendThisMonth = expense,
                    totalIncomeThisMonth = income,
                    balance = balance
                )
            }
        }
    }

    fun refresh(userId: String) {
        loadRecentTransactions(userId)
        loadMonthlyTotal(userId)
    }
}

File: ios/BudgetBuddy/ViewModels/HomeViewModel.swift
import Foundation
import Combine

struct HomeUiState {
    var balance: Double               = 0
    var totalIncomeThisMonth: Double  = 0
    var totalSpendThisMonth: Double   = 0
    var allTimeIncome: Double         = 0
    var allTimeExpense: Double        = 0
    var recentTransactions: [BBTransaction] = []
    var activeGoals: [BBGoal]         = []
    var categories: [BBCategory]      = []
    var isLoading: Bool               = false
    var error: String?                = nil
}

@MainActor
final class HomeViewModel: ObservableObject {
    @Published var uiState = HomeUiState()

    private let transactionRepo: TransactionRepository
    private let budgetRepo: BudgetRepository
    private let goalRepo: GoalRepository
    private let categoryRepo: CategoryRepository

    init(transactionRepo: TransactionRepository, budgetRepo: BudgetRepository,
         goalRepo: GoalRepository, categoryRepo: CategoryRepository) {
        self.transactionRepo = transactionRepo
        self.budgetRepo      = budgetRepo
        self.goalRepo        = goalRepo
        self.categoryRepo    = categoryRepo
    }

    func load(userId: String) {
        uiState.isLoading = true
        Task {
            do {
                let categories = try categoryRepo.all()
                let recent     = try transactionRepo.recent(userId: userId, limit: 5)
                let goals      = try goalRepo.active(userId: userId)

                let start      = DateUtils.startOfMonth()
                let end        = DateUtils.endOfMonth()
                let income     = try transactionRepo.totalIncome(userId: userId, start: start, end: end)
                let expense    = try transactionRepo.totalExpense(userId: userId, start: start, end: end)

                let allTxs     = try transactionRepo.all(userId: userId)
                let allIncome  = allTxs.filter { $0.type == .income  }.reduce(0) { $0 + $1.amount }
                let allExpense = allTxs.filter { $0.type == .expense }.reduce(0) { $0 + $1.amount }

                uiState = HomeUiState(
                    balance:               income - expense,
                    totalIncomeThisMonth:  income,
                    totalSpendThisMonth:   expense,
                    allTimeIncome:         allIncome,
                    allTimeExpense:        allExpense,
                    recentTransactions:    recent,
                    activeGoals:           goals,
                    categories:            categories,
                    isLoading:             false
                )
            } catch {
                uiState.isLoading = false
                uiState.error     = error.localizedDescription
            }
        }
    }
}
